[5.0.3 regression] error: "type[RealmFilter]" has no attribute "objects"  [attr-defined]

When trying to upgrade to django-stubs 5.0.3 or 5.0.4 in [Zulip](https://github.com/zulip/zulip), I get 25 unexpected errors like `error: "type[RealmFilter]" has no attribute "objects"  [attr-defined]`. This is not #1684; `RealmFilter` is a normal concrete model class. The issue seems to depend on the structure of the project’s import graph.

Based on `git bisect`, it first occurs at

- 2f5a4946e0569e9bc69729fc59794a1ea4347b91 (#2276, cc @asottile).

I’ve reduced the issue to this minimized test case: https://github.com/andersk/no-attribute-objects

```console
$ mypy .
zerver/models/__init__.py:6: error: "type[RealmFilter]" has no attribute "objects"  [attr-defined]
Found 1 error in 1 file (checked 10 source files)
```

**`pyproject.toml`**

```toml
[tool.django-stubs]
django_settings_module = "zproject.settings"

[tool.mypy]
plugins = ["mypy_django_plugin.main"]
```

**`confirmation/__init__.py`** (empty)

**`confirmation/models.py`**

```python
from django.db import models

from zerver.models import Realm


class Confirmation(models.Model):
    realm = models.ForeignKey(Realm, on_delete=models.RESTRICT)
```

**`zerver/__init__.py`** (empty)

**`zerver/models/__init__.py`**

```python
from zerver.models.linkifiers import RealmFilter as RealmFilter
from zerver.models.realms import Realm as Realm
from zerver.models.streams import Stream as Stream
from zerver.models.users import UserProfile as UserProfile

RealmFilter.objects
```

**`zerver/models/linkifiers.py`**

```python
from django.db import models


class RealmFilter(models.Model):
    pass
```

**`zerver/models/realms.py`**

```python
from django.db import models


class Realm(models.Model):
    pass
```

**`zerver/models/streams.py`**

```python
from django.db import models

from zerver.models.realms import Realm
from zerver.models.users import UserProfile


class Stream(models.Model):
    realm = models.ForeignKey(Realm, on_delete=models.RESTRICT)
    creator = models.ForeignKey(UserProfile, on_delete=models.RESTRICT)
```

**`zerver/models/users.py`**

```python
from django.contrib.auth.models import AbstractBaseUser, PermissionsMixin


class UserProfile(AbstractBaseUser, PermissionsMixin):
    pass
```

**`zproject/__init__.py`** (empty)

**`zproject/settings.py`**

```python
INSTALLED_APPS = [
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "confirmation",
    "zerver",
]
```

--- a/mypy_django_plugin/main.py
+++ b/mypy_django_plugin/main.py
@@ -223,9 +223,6 @@ class NewSemanalDjangoPlugin(Plugin):
         return None
 
     def get_customize_class_mro_hook(self, fullname: str) -> Optional[Callable[[ClassDefContext], None]]:
-        if fullname == fullnames.MODEL_CLASS_FULLNAME:
-            return MetaclassAdjustments.adjust_model_class
-
         sym = self.lookup_fully_qualified(fullname)
         if (
             sym is not None
@@ -235,6 +232,11 @@ class NewSemanalDjangoPlugin(Plugin):
             return reparametrize_any_manager_hook
         else:
             return None
+
+    def get_metaclass_hook(self, fullname: str) -> Optional[Callable[[ClassDefContext], None]]:
+        if fullname == fullnames.MODEL_METACLASS_FULLNAME:
+            return MetaclassAdjustments.adjust_model_class
+        return None
 
     def get_base_class_hook(self, fullname: str) -> Optional[Callable[[ClassDefContext], None]]:
         # Base class is a Model class definition

--- a/tests/typecheck/managers/test_managers.yml
+++ b/tests/typecheck/managers/test_managers.yml
@@ -761,3 +761,53 @@
                     generic_manager = models.Manager()
                     generic_manager_from_generic_queryset = GenericManagerFromGenericQuerySet()
                     generic_manager_from_populated_queryset = GenericManagerFromPopulatedQuerySet()
+
+# Regression test for #2304
+-   case: test_objects_managers_is_kept_with_specific_import_graph
+    main: |
+        from zerver.models import RealmFilter
+        reveal_type(RealmFilter.objects)  # N: Revealed type is "django.db.models.manager.Manager[zerver.models.linkifiers.RealmFilter]"
+    installed_apps:
+        - django.contrib.auth
+        - django.contrib.contenttypes
+        - confirmation
+        - zerver
+    files:
+        -   path: confirmation/__init__.py
+        -   path: confirmation/models.py
+            content: |
+                from django.db import models
+                from zerver.models import Realm
+                class Confirmation(models.Model):
+                    realm = models.ForeignKey(Realm, on_delete=models.CASCADE)
+        -   path: zerver/__init__.py
+        -   path: zerver/models/__init__.py
+            content: |
+                from zerver.models.linkifiers import RealmFilter as RealmFilter
+                from zerver.models.realms import Realm as Realm
+                from zerver.models.streams import Stream as Stream
+                from zerver.models.users import UserProfile as UserProfile
+                RealmFilter.objects
+        -   path: zerver/models/linkifiers.py
+            content: |
+                from django.db import models
+                class RealmFilter(models.Model):
+                    pass
+        -   path: zerver/models/realms.py
+            content: |
+                from django.db import models
+                class Realm(models.Model):
+                    pass
+        -   path: zerver/models/streams.py
+            content: |
+                from django.db import models
+                from zerver.models.realms import Realm
+                from zerver.models.users import UserProfile
+                class Stream(models.Model):
+                    realm = models.ForeignKey(Realm, on_delete=models.RESTRICT)
+                    creator = models.ForeignKey(UserProfile, on_delete=models.RESTRICT)
+        -   path: zerver/models/users.py
+            content: |
+                from django.contrib.auth.models import AbstractBaseUser, PermissionsMixin
+                class UserProfile(AbstractBaseUser, PermissionsMixin):
+                    pass

